### ISSUE TITLE
Quarkus 2.9.x regression with Panache/Hibernate autoincrementing tables when upgrading from 2.8.3

### ISSUE THREAD

**Issue by kenyee** [quarkusio/quarkus#25958]
### Describe the bug

Hibernate persistence seems to be setting the ID to an incorrect value when persisting a new object in 2.9.x (I've seen it in version 2.9.0-2.9.2 but finally had some time to create a repro sample).
The database is configured to autogenerate the ID, so Hibernate should not be setting it (leaving it null will let the DB autogenerate the ID).  It's also set to an incorrect value (1) which collides with data in the database.

This is different behavior from 2.8.3.
I've attached a repro.  Run the only test in it and you'll see the error.  Downgrade to 2.8.3 and you won't see failure.

### Expected behavior

Hibernate shouldn't be setting object IDs when inserting new objects into a table that has autoincrementing values.

### Actual behavior

You'll see this in the logs:
```
****Count is 2
Hibernate: insert into Project (id, name, retentionMonths) values (default, ?, ?)
Jun 03, 2022 4:56:59 PM org.hibernate.engine.jdbc.spi.SqlExceptionHelper logExceptions
WARN: SQL Error: 23505, SQLState: 23505
Jun 03, 2022 4:56:59 PM org.hibernate.engine.jdbc.spi.SqlExceptionHelper logExceptions
ERROR: Unique index or primary key violation: "PRIMARY KEY ON PUBLIC.PROJECT(ID) ( /* key:1 */ CAST(1 AS BIGINT), 'initial-project', 60)"; SQL statement:
insert into Project (id, name, retentionMonths) values (default, ?, ?) [23505-210]
```

### How to Reproduce?

See attached project.
Run the  only test in the project using IntelliJ or do "./gradlew check" from the command line.
[bugquarkusjpatest.zip](https://github.com/quarkusio/quarkus/files/8835703/bugquarkusjpatest.zip)

### Output of `uname -a` or `ver`

Darwin tw-mbp-keny 21.5.0 Darwin Kernel Version 21.5.0: Tue Apr 26 21:08:22 PDT 2022; root:xnu-8020.121.3~4/RELEASE_X86_64 x86_64

### Output of `java -version`

openjdk version "11.0.13" 2021-10-19 OpenJDK Runtime Environment GraalVM CE 21.3.0 (build 11.0.13+7-jvmci-21.3-b05) OpenJDK 64-Bit Server VM GraalVM CE 21.3.0 (build 11.0.13+7-jvmci-21.3-b05, mixed mode, sharing)

### GraalVM version (if different from Java)

21.3.0

### Quarkus version or git rev

2.9.2-Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

 ------------------------------------------------------------ Gradle 7.4.2 ------------------------------------------------------------  Build time:   2022-03-31 15:25:29 UTC Revision:     540473b8118064efcc264694cbcaa4b677f61041  Kotlin:       1.5.31 Groovy:       3.0.9 Ant:          Apache Ant(TM) version 1.10.11 compiled on July 10 2021 JVM:          11.0.13 (GraalVM Community 11.0.13+7-jvmci-21.3-b05) OS:           Mac OS X 12.4 x86_64

**Comment by gsmet** [comment]
@beikov could you check if this one is a problem with Hibernate ORM?

**Comment by beikov** [comment]
As far as I understand the test case, you are creating a table with `auto_increment`, but then you persist a row with an explicit identifier value. The database does not know that it has to increment/change the internal sequence though, so when Hibernate comes along and tries to insert into the database, it will let the database do it's work and use the `auto_increment` feature of the database.

In short, the problem is with how you use this, not with Hibernate or anything else. If you want to insert something into the database with a well known identifier, you will have to adjust the internal sequence value of the table afterwards to avoid collisions. In MySQL, you can use `ALTER TABLE table_name AUTO_INCREMENT = start_value;` for this purpose.

**Comment by kenyee** [comment]
Hi @beikov,
Actually, when persisted by my code, the ID is not set at all:
```
    fun testReadWriteProject() {
        val project = Project().apply {
            name = "test"
        }
        val oldCount = projectRepository.count()
        projectRepository.persist(project)
```

From the logs, it does look like it's set, which is why I thought it was a Hibernate bug.
The SQL schema is set up for the DB to handle the autoincrementing:
```
create table Project (
    id bigint not null auto_increment primary key,
```

and Hibernate should be letting the persistence provider generate identity values:
```
class Project {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
```

Doesn't GenerationType.IDENTITY mean "use an autoincrementing column on the table schema" and don't set the ID during the insert row process?  When I tell Hibernate to create the DB schema, it generates an autoincrement column.
The .SEQUENCE and .TABLE generation types mean that Hibernate will store the sequence in a table and increment as I understood it (and will also send the ID during the insert of a new row).

**Comment by beikov** [comment]
>Actually, when persisted by my code, the ID is not set at all:

This might be a Kotlin issue then, though I don't know what you mean by "not set at all". I mostly focused on the SQL part, and in the SQL you can see that Hibernate emits the `default` keyword in `insert into project(id, name) values (default, ?)` which means the identity generation works fine. The test case fails with a constraint violation though, because your liquibase script does an insert with a fixed value.

This is a common mistake with identity generation and other databases allow you to overcome these situations by forcing developers to always use `default`, when using the `generated always as identity` clause i.e. essentially forbidding to set a custom value.

**Comment by kenyee** [comment]
> I mostly focused on the SQL part, and in the SQL you can see that Hibernate emits the default keyword in insert into project(id, name) values (default, ?) which means the identity generation works fine. The test case fails with a constraint violation though, because your liquibase script does an insert with a fixed value.

Hmm..I actually wonder if this is a bug in the H2 hibernate provider instead.
So the sequence is actually this:
1) liquibase creates DB schema
2) liquibase initializes DB w/ some initial data (data has IDs 1 and 2)
3) test runs (the count = 2 is printed at the start of the test because I wanted to confirm there was no race condition with part)
4) the test run this command:
Hibernate: insert into Project (id, name, retentionMonths) values (default, ?, ?)
and "default" is somehow set to "1" during the insert. 🤔 

Switching to MySQL hibernate provider works fine and this coincided with the upgrade in Quarkus to the latest version H2 in Quarkus 2.9...
Makes no sense for this to be a Kotlin bug IMHO...

**Comment by beikov** [comment]
Your database believes the next value to use for "auto_increment" is 1, because it didn't advance this internal sequence yet. Change your insert script in your liquibase changelog to also use `default` instead of a fixed value `1`, then the database will increment its internal sequence value.

**Comment by beikov** [comment]
I have no idea why MySQL doesn't fail with the same error. Maybe it does increment the internal sequence automatically somehow even when a normal insert is used, but this is a total non-standard feature.

The thing about the possible Kotlin issue was, that you might not see the value on the entity object due to some accessor or mutability magic Kotlin does behind the scenes. I don't know Kotlin well enough, but this is why I asked you to tell me what you mean by "not set at all".

Either way, it appears to me that the root of the issue is simply the fact that the PK value which the auto increment sequence returns is already taken, because the sequence is not used/updated.

**Comment by kenyee** [comment]
> Change your insert script in your liquibase changelog to also use default instead of a fixed value 1

That was it.  Thanks for solving that mystery 😄 
It was indeed using fixed values for the insert....apparently the new H2 database version requires using "default" instead.

**Comment by dvdmatt** [comment]
This replicated in Quarkus quarkus:3.6.4

matt@vm008:247% java --version
java 17.0.9 2023-10-17 LTS
Java(TM) SE Runtime Environment Oracle GraalVM 17.0.9+11.1 (build 17.0.9+11-LTS-jvmci-23.0-b21)
Java HotSpot(TM) 64-Bit Server VM Oracle GraalVM 17.0.9+11.1 (build 17.0.9+11-LTS-jvmci-23.0-b21, mixed mode, sharing)
matt@vm008:248% native-image --version
native-image 17.0.9 2023-10-17
GraalVM Runtime Environment Oracle GraalVM 17.0.9+11.1 (build 17.0.9+11-LTS-jvmci-23.0-b21)
Substrate VM Oracle GraalVM 17.0.9+11.1 (build 17.0.9+11-LTS, serial gc, compressed references)

<code>
        LOGGER.info("Creating new module " + module);
        return module.<Module>persist();
</code>

<log>
22:15:53 INFO  [co.li.ca.ne.mo.ModuleResource] (vert.x-eventloop-thread-1) Creating new module {"id":null,"playerId":0,"name":"AAA","moduleType":"Projectile","technology":10,"size":10}
22:15:53 WARN  [or.hi.en.jd.sp.SqlExceptionHelper] (vert.x-eventloop-thread-1) SQL Error: 1062, SQLState: 23000
22:15:53 ERROR [or.hi.en.jd.sp.SqlExceptionHelper] (vert.x-eventloop-thread-1) {errorMessage=Duplicate entry '4' for key 'PRIMARY', errorCode=1062, sqlState=23000}
22:15:53 ERROR [io.qu.ve.ht.ru.QuarkusErrorHandler] (vert.x-eventloop-thread-1) HTTP Request to /nebula/module failed, error id: 41b76185-f773-4dfe-9c60-3aec10177b6f-3: org.hibernate.exception.ConstraintViolationException: error executing SQL statement [{errorMessage=Duplicate entry '4' for key 'PRIMARY', errorCode=1062, sqlState=23000}] [insert into Module (moduleType,name,playerId,size,technology,id) values (?,?,?,?,?,?)]
</log>

What additional detail would help?

**Comment by FroMage** [comment]
Can you open a new issue for this please?
